### ISSUE TITLE
SelectMany unable to  unnest an array of strings

### ISSUE THREAD

**Issue by calhamdower** [couchbaselabs/Linq2Couchbase#171]
Not Sure if there's meant to be support for SelectMany? Its not really detailed in the docs here that I could find.

Db.Query<Cat>().SelectMany(a => a.Feet).Explain();

I found though even running the Explain N1QL in CBQ wasnt a proper response:
"SELECT `Extent2`.* FROM `Testing` as `Extent1` INNER UNNEST `Extent1`.`teamList` as `Extent2` WHERE (`Extent1`.`type` = 'Sport')"
results in:
[
  {},
  {},
  {},
  {}
]
Maybe it should result in "SELECT `Extent2` as `result` FR......"

Couchbase.Linq.CouchbaseQueryException occurred
  HResult=-2146232832
  Message=An error occurred executing the N1QL query.  See the inner exception for details.
  Source=Couchbase.Linq
   InnerException: 
       HResult=-2146233088
       LineNumber=8
       LinePosition=9
       Message=Unexpected character encountered while parsing value: {. Path 'results', line 8, position 9.
       Path=results
       Source=Newtonsoft.Json

**Comment by brantburnett** [comment]
@calhamdower 

Looking at it, I think the generated query looks correct.  But it's incorrectly trying to unwrap it from a results object.

Can you do a couple of things that will help me diagnose.

1. Can you provide a couple of sample documents and your POCOs?
2. Can you try running the query this way to see if it has different behavior?  It might build the lambdas differently:

```csharp
from p in Db.Query()
from q in p.Feet
select q
```

Thanks,
Brant

**Comment by calhamdower** [comment]
@brantburnett 
Yes I tried that one earlier thinking it might resolve differently... though both:
var ids = (from p in Db.Query<Sport>()
                from q in p.TeamList
                select q).toList();
var ids2 =DAL.CbDal.BucketContext.Query<Sport>().Where(a=>a.Type=="Sport").SelectMany(a =>TeamList).ToList();
Result in the exact same fault. 

Fault occurs when the linq expression is evaluated via toList etc... so I think you're right in that it can't unwrap the result. when I run the N1QL expression in cbq however the result doesn't look as I'd expect also... I'd expect the result to appear as: 
```
"result": [
      "Team::9fb8b28813ea4132b8e8ce4f86c1664c",
      "Team::6012cf5cb6c84830a1a162157c1d3ad8"
    ]
rather than 
[
{},
{},
{},
{}
]
```
Which may be the cause of the inability to Parse the result? 

POCO:
```
public class Sport
{
        [JsonProperty("type")]
        public string Type => GetType().Name;
        [JsonProperty("id")]
        public string Id { get; set; }
        [JsonProperty("sportName")]
        public string SportName { get; set; }
        [JsonProperty("teamList")]
        public List<string> TeamList { get; set; }
        [JsonProperty("annModelList")]
        public List<string> AnnModelList { get; set; }
}

```

Thanks 

[SampleSports.txt](https://github.com/couchbaselabs/Linq2Couchbase/files/659880/SampleSports.txt)

**Comment by brantburnett** [comment]
@calhamdower 

Okay, I see what's going on now.  The issue is that you're unnesting an array of strings rather than an array of objects, which is an edge case we apparently missed.  The generated query should be:

```sql
SELECT `Extent2` as `result` FROM `Testing` as `Extent1` INNER UNNEST `Extent1`.`teamList` as `Extent2` WHERE (`Extent1`.`type` = 'Sport')
```

This would then result in this query result:

```json
[
{
  "result": "Team::9fb8b28813ea4132b8e8ce4f86c1664c"
},
{
  "result": "Team::6012cf5cb6c84830a1a162157c1d3ad8"
}
]
```

With that result, the deserializer would then be able to interpret it correctly.  Let me see what I can do to get this fixed.

In the meantime, you should be able to work around this problem like this:

```csharp
DAL.CbDal.BucketContext.Query().Where(a=>a.Type=="Sport").AsEnumerable().SelectMany(a =>TeamList).ToList()
```

The added call to `AsEnumerable` will trigger the query before the SelectMany call, and then execute SelectMany in memory on the client.  It isn't as efficient because it's pulling back the whole object, but it should work for you until we get a fix deployed.

Brant

**Comment by calhamdower** [comment]
@brantburnett Thanks for the Time/Analysis Brant. 

This code you suggested is fine for me at the moment as my "sport" documents are not too large anyway so there's not much perf hits. 

Good to know I wasn't just being a pleb.

Ill follow Issue #171 to see what release the fix might come in, much appreciated.

Calham

**Comment by brantburnett** [comment]
@jeffrymorris 

This can be closed, was addressed by pull #175
